circle/src/domain.rs

```diff
@@ -41,59 +41,55 @@ use crate::util::{point_to_univariate, s_p_at_p, univariate_to_point, v_0, v_n};
+            shift: Point::generator(log_n + 1),
         }
     }
     fn is_standard(&self) -> bool {
-        self.shift == F::circle_two_adic_generator(self.log_n + 1)
+        self.shift == Point::generator(self.log_n + 1)
     }
-    pub(crate) fn points(&self) -> impl Iterator<Item = Complex<F>> {
-        let half_gen = F::circle_two_adic_generator(self.log_n - 1);
-        let coset0 = half_gen.shifted_powers(self.shift);
-        let coset1 = half_gen.shifted_powers(half_gen / self.shift);
-        coset0.interleave(coset1).take(1 << self.log_n)
+    pub(crate) fn gen(&self) -> Point<F> {
+        Point::generator(self.log_n - 1)
     }
-    pub(crate) fn nth_point(&self, idx: usize) -> Complex<F> {
-        // Only implemented for standard position.
-        assert!(self.is_standard());
-        let gen = F::circle_two_adic_generator(self.log_n);
-        self.shift * gen.exp_u64(idx as u64)
+    pub(crate) fn coset0(&self) -> impl Iterator<Item = Point<F>> {
+        let g = self.gen();
+        iterate(self.shift, move |&p| p + g).take(1 << (self.log_n - 1))
+    }
+    fn coset1(&self) -> impl Iterator<Item = Point<F>> {
+        let g = self.gen();
+        iterate(g - self.shift, move |&p| p + g).take(1 << (self.log_n - 1))
+    }
+    pub(crate) fn points(&self) -> impl Iterator<Item = Point<F>> {
+        self.coset0().interleave(self.coset1())
+    }
+    pub(crate) fn nth_point(&self, idx: usize) -> Point<F> {
+        let (idx, lsb) = (idx >> 1, idx & 1);
+        if lsb == 0 {
+            self.shift + self.gen() * idx
+        } else {
+            -self.shift + self.gen() * (idx + 1)
+        }
     }
 
-    /// Computes the lagrange basis at point, not yet normalized by the value of the domain
-    /// vanishing poly, since that is more efficient to compute after the dot product.
-    #[instrument(skip_all, fields(log_n = %self.log_n))]
-    pub(crate) fn lagrange_basis<EF: ExtensionField<F>>(&self, point: Complex<EF>) -> Vec<EF> {
-        let domain = self.points().collect_vec();
-
-        // the denominator so that the lagrange basis is normalized to 1
-        // TODO: this depends only on domain, so should be precomputed
-        let lagrange_normalizer: Vec<F> = domain
-            .iter()
-            .map(|p| s_p_at_p(p.real(), p.imag(), self.log_n))
-            .collect();
+    pub(crate) fn zeroifier<EF: ExtensionField<F>>(&self, at: Point<EF>) -> EF {
+        at.v_n(self.log_n) - self.shift.v_n(self.log_n)
+    }
 
-        let basis = domain
-            .into_iter()
-            .zip(&lagrange_normalizer)
-            .map(|(p, &ln)| {
-                // ext * base
-                // TODO: this can be sped up
-                v_0(p.conjugate().rotate(point)) * ln
-            })
-            .collect_vec();
+    pub(crate) fn s_p<EF: ExtensionField<F>>(&self, p: Point<F>, at: Point<EF>) -> EF {
+        self.zeroifier(at) / p.v_tilde_p(at)
+    }
```

Comment:
Maybe worth to comment that this is the non-normalized single-point selector / Lagrangian.